Go back to comparing timestamps in has_database_updated_with_new_score

efischer19 opened the pull request:
WIP, commiting in discrete chunks

efischer19 commented on lms/djangoapps/grades/tasks.py:
```diff
@@ -24,9 +26,26 @@
+    )
+
+@task(default_retry_delay=30, routing_key=settings.RECALCULATE_GRADES_ROUTING_KEY)
+def _recalculate_subsection_grade(**kwargs):
```
@jibsheet thoughts on this method of collapsing our task signature down to a single `**kwargs`? We're going to be removing `weighted_earned` and `weighted_possible` anyways, so I figure why not go whole-hog.

jibsheet replied:
Is the thought move to _ for a release then move back?  I think that's the only real way in the current queue world.

efischer19 replied:
Yeah, that's what I was planning. After that though, we'll only have the single kwargs argument, so future changes will be easier and we shouldn't have to do this repeatedly.

jibsheet replied:
Great - I agree that's the 'best' pattern .

efischer19 commented:
The timestamp work, https://openedx.atlassian.net/browse/TNL-5994 is now done. Moving on to the ORA work, https://openedx.atlassian.net/browse/TNL-5995

efischer19 commented on lms/djangoapps/grades/tasks.py:
```diff
@@ -36,63 +56,59 @@ def recalculate_subsection_grade(
     expected new score values for the given problem and user.
     """
-    score = get_score(user_id, scored_block_usage_key)
+    score = get_full_score(user_id, scored_block_usage_key)
```
On today's episode of "Boy, that was lucky" - turns out we don't need to do anything special for ORA problems? The new `get_full_score` method returns a `StudentModule` object. As it turns out, scores that come from submissions *do* have an entry in that table, the problem was just that it didn't define 'grade' or 'max_grade'. It has a 'modified' value though, so the new code works fine.

efischer19 replied:
Damn, nvm. Turns out that `modified` value isn't always updated when a new submissions score is set. Back to the drawing board...

efischer19 commented:
@nasthagiri @sanfordstudent @jcdyer @yro In the interest of merging smaller PRs, let's go ahead and review/merge this work for https://openedx.atlassian.net/browse/TNL-5994 independently while I keep digging into the ORA issue. Any volunteers for review?

efischer19 commented on lms/djangoapps/grades/tasks.py:
```diff
@@ -24,9 +26,27 @@
+            course_id=course_id,
+            usage_id=usage_id,
+            only_if_higher=only_if_higher,
+            task_queued_time=kwargs.get('task_queued_time', 0),  # If no time, assume 1/1/1970
```
This is a leftover remnant from when I was toying with using unix timestamps, one second...

efischer19 commented:
jenkins run js

nasthagiri commented on lms/djangoapps/grades/signals/handlers.py:
```diff
@@ -158,14 +159,13 @@ def enqueue_subsection_update(sender, **kwargs):  # pylint: disable=unused-argum
     enqueueing a subsection update operation to occur asynchronously.
     """
-    result = recalculate_subsection_grade.apply_async(
+    result = _recalculate_subsection_grade.apply_async(
```
Why was this converted to an internal method name when it's used in non-test code outside of its python module?

efischer19 replied:
It's part of the temporary shim to clean up the `recalculate_subsection_grade` handler. The new, correct method name is temporarily being labelled `_name` for this release, so that pre-existing tasks queued against `name` can still be run. However, new code can call the correct new method, and does so here.

The intent is to remove the shim and rename the new method to be just `recalculate_subsection_grade` (with a single `**kwargs` in the method signature) over the next 2 releases.

nasthagiri commented on lms/djangoapps/grades/signals/handlers.py:
```diff
@@ -158,14 +159,13 @@ def enqueue_subsection_update(sender, **kwargs):  # pylint: disable=unused-argum
             only_if_higher=kwargs.get('only_if_higher'),
-            weighted_earned=kwargs.get('weighted_earned'),
-            weighted_possible=kwargs.get('weighted_possible'),
+            task_queued_time=unicode(datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S.%f")),
```
Please create a helper for getting the current time as a string in util/date_utils.py.

nasthagiri commented on lms/djangoapps/grades/tasks.py:
```diff
@@ -24,9 +26,27 @@
         user_id, course_id, usage_id, only_if_higher, weighted_earned, weighted_possible, **kwargs
 ):
     """
+    Shim to allow us to modify this task's signature without blowing up production on deployment.
```
It's still going to blow up since the new code expects the new kwargs to be there.

efischer19 replied:
I think we should be fine - previously queued tasks will have an arguments list of `user_id, course_id, usage_id, only_if_higher, weighted_earned, weighted_possible, **kwargs`. From that, we drop `weighted_earned` and `weighted_possible`, and shim in a dummy value for `task_queued_time`, which allows us to call the new method signature cleanly.

nasthagiri replied:
That's true.  But new LMS instances could still be queuing onto old celery workers during deployment.  The old celery workers will be expecting the old kwargs.

efischer19 replied:
That's why all instances that were previously calling `recalculate_subsection_grades.apply_async(...)` (no underscore) have been changed to the shiny new `_recalculate_subsection_grades.apply_async(...)` (yes underscore). All newly queued tasks will be using the new method signature, no different than if we had added a totally new task.

nasthagiri replied:
ah.  that makes sense.  great idea.
We should still rename it though since it's not an underscore-like function.  You could add a version number instead like recalculate_subsection_grades_v2.  And if this design pattern works, we can keep incrementing the versions (and delete old versions once newer ones are deployed).

nasthagiri commented on lms/djangoapps/grades/tasks.py:
```diff
@@ -36,63 +56,57 @@ def recalculate_subsection_grade(
     score_deleted = kwargs['score_deleted']
-    scored_block_usage_key = UsageKey.from_string(usage_id).replace(course_key=course_key)
+    scored_block_usage_key = UsageKey.from_string(kwargs['usage_id']).replace(course_key=course_key)
+    task_queued_time = datetime.strptime(kwargs['task_queued_time'], "%Y-%m-%d %H:%M:%S.%f").replace(tzinfo=pytz.UTC)
```
Please create a common helper for converting from string to datetime object in date_utils.py.

nasthagiri commented on lms/djangoapps/grades/tasks.py:
```diff
@@ -36,63 +56,57 @@ def recalculate_subsection_grade(
-        found_raw_possible == expected_raw_possible
-    )
+    # In non-async environments, we're still inside an uncommited transaction. This means that score.modified will be
+    # reporting a time slightly earlier than task_queued_time, so we add a "fudge factor" of 1/10 of a second.
```
Rather than addressing it this way, I suggest passing the `modified` value through rather than computing the current timestamp.  This would:
1. avoid atomic time issues.
2. work for both async and sync operations.

nasthagiri commented:
I believe it will be better to figure out how to get this working with ORA before we merge this PR.  Given that we will be breaking the interface again.

efischer19 commented on requirements/edx/github.txt:
```diff
@@ -76,8 +76,8 @@ git+https://github.com/edx/XBlock.git@xblock-0.4.12#egg=XBlock==0.4.12
-git+https://github.com/edx/edx-ora2.git@1.1.11#egg=ora2==1.1.11
--e git+https://github.com/edx/edx-submissions.git@1.1.1#egg=edx-submissions==1.1.1
+git+https://github.com/edx/edx-ora2.git@1.1.12#egg=ora2==1.1.12
+-e git+https://github.com/edx/edx-submissions.git@1.1.4#egg=edx-submissions==1.1.4
```
Neither of these versions actually exist yet, so expect to see some test failures in the short-term

efischer19 commented on lms/djangoapps/grades/tasks.py:
```diff
@@ -24,9 +28,27 @@
+
+
+@task(default_retry_delay=30, routing_key=settings.RECALCULATE_GRADES_ROUTING_KEY)
+def recalculate_subsection_grade_v2(**kwargs):
```
@jibsheet This is the change we've discussed to prevent blowing up worker error reporting while deploying. I'd almost rather avoid the `_vN` in the signature and just do a one-time shim to get us to a clean `recalculate_subsection_grade(**kwargs)` which can be kept backwards compatible internally, how do you feel?

nasthagiri replied:
Remember that just having `kwargs` in the signature doesn't allow the task to be backward and forward compatible.  Task workers will expect certain arguments to be there for correctness of the feature.  Similarly, enqueuers will expect that their task is being handled by the matching version of the code.

So versioning these tasks are the most full-proof way of handling multi-version deployments.  This assumes that old workers who cannot handle new-versioned tasks will ignore those tasks and leave them for new workers.

nasthagiri replied:
Old -> New: New needs to handle the case where the parameter doesn't exist.
New -> Old: New needs to handle the case where the parameter won't be honored.

We version the task so that the latter does not happen.

efischer19 commented:
@nasthagiri @sanfordstudent This is ready for a round of review

nasthagiri commented on lms/djangoapps/grades/tasks.py:
```diff
@@ -87,12 +104,24 @@ def _has_database_updated_with_new_score(
         # score should be None only if it was deleted.
         # Otherwise, it hasn't yet been saved.
         return score_deleted
+    elif score.grade is None:  # submissions-related problem
```
Can you check the `module_type` instead?  I'm not sure that checking for `grade` will always work.

efischer19 commented:
I've rebased/handled merge conflicts and added a few very minor test fixes (mock signals in other modules that weren't updated with the `modified` value).

efischer19 commented on lms/djangoapps/courseware/model_data.py:
```diff
@@ -1024,4 +1025,4 @@ def get_score(user_id, usage_key):
         return None
     else:
-        return student_module.grade, student_module.max_grade
+        return student_module
```
It contains all of these things; most of which are, strictly speaking, a superset of the data specifically related to "scores". However, `set_score` above is still directly calling `StudentModule.objects.get_or_create()`, so the inconsistency in referring to a StudentModule object as a "score" is consistent between the two methods.

```
(Pdb) dir(student_module)
['DONE_TYPES', 'DoesNotExist', 'MODEL_TAGS', 'MODULE_TYPES', 'MultipleObjectsReturned', '__class__', '__delattr__', '__dict__', '__doc__', '__eq__', '__format__', '__getattribute__', '__hash__', '__init__', u'__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setstate__', '__sizeof__', '__str__', '__subclasshook__', '__unicode__', '__weakref__', '_base_manager', '_check_column_name_clashes', '_check_field_name_clashes', '_check_fields', '_check_id_field', '_check_index_together', '_check_local_fields', '_check_long_column_names', '_check_m2m_through_same_relationship', '_check_managers', '_check_model', '_check_ordering', '_check_swappable', '_check_unique_together', '_default_manager', '_deferred', '_do_insert', '_do_update', '_get_FIELD_display', '_get_next_or_previous_by_FIELD', '_get_next_or_previous_in_order', '_get_pk_val', '_get_unique_checks', '_meta', '_perform_date_checks', '_perform_unique_checks', '_save_parents', '_save_table', '_set_pk_val', '_state', 'all_submitted_problems_read_only', 'check', 'clean', 'clean_fields', 'course_id', 'created', 'date_error_message', 'delete', 'done', 'from_db', 'full_clean', 'get_deferred_fields', 'get_done_display', 'get_module_type_display', 'get_next_by_created', 'get_next_by_modified', 'get_previous_by_created', 'get_previous_by_modified', 'grade', 'id', 'max_grade', 'modified', 'module_state_key', 'module_type', 'objects', 'pk', 'prepare_database_save', 'refresh_from_db', 'save', 'save_base', 'serializable_value', 'state', 'student', 'student_id', 'studentmodulehistory_set', 'studentmodulehistoryextended_set', 'unique_error_message', 'validate_unique']
```

efischer19 commented:
There's a lettuce test here that keeps failing and seems like it may actually be related, as it's related to LTI grading: https://build.testeng.edx.org/job/edx-platform-test-subset/676230/

I'll still looking

efischer19 commented on common/lib/xmodule/xmodule/lti_2_util.py:
```diff
@@ -231,7 +231,7 @@ def clear_user_module_score(self, user):
             nothing
         """
-        self.set_user_module_score(user, None, None)
+        self.set_user_module_score(user, 0, self.max_score())
```
This is one of the dumbest and most confusing test failures I've seen in a while. This call is trying to set a score of `(None, None)`, and then after passing through the grade signal handlers, lti complains that the values of possible and earned cannot be None.

```
    And the LTI provider deletes my grade and feedback                                              # lms/djangoapps/courseware/features/lti.py:367
ERROR:courseware.module_render:HEY LOOK AT THIS
ERROR:courseware.module_render:['max_value', 'user_id', 'value']
ERROR:lms.djangoapps.grades.signals.handlers:EFISCHER DEBUG SPECIAL!
ERROR:lms.djangoapps.grades.signals.handlers:kwargs: {{'only_if_higher': None, 'user_id': 123, 'weight': 10.0, 'signal': <django.dispatch.dispatcher.Signal object at 0x7f319e9cbf50>, 'modified': datetime.datetime(2016, 12, 8, 21, 4, 38, 203510, tzinfo=<UTC>), 'course_id': u'edx/test_course_a/Test_Course', 'raw_possible': None, 'raw_earned': None, 'usage_id': u'i4x://edx/test_course_a/lti/LTI'}}
  File "/usr/lib/python2.7/threading.py", line 783, in __bootstrap
    self.__bootstrap_inner()
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/lib/python2.7/SocketServer.py", line 599, in process_request_thread
    self.finish_request(request, client_address)
  File "/usr/lib/python2.7/SocketServer.py", line 334, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/servers/basehttp.py", line 102, in __init__
    super(WSGIRequestHandler, self).__init__(*args, **kwargs)
  File "/usr/lib/python2.7/SocketServer.py", line 655, in __init__
    self.handle()
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/lettuce/django/server.py", line 95, in handle
    handler.run(self.server.get_app())
  File "/usr/lib/python2.7/wsgiref/handlers.py", line 85, in run
    self.result = application(self.environ, self.start_response)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/contrib/staticfiles/handlers.py", line 63, in __call__
    return self.application(environ, start_response)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/wsgi.py", line 189, in __call__
    response = self.get_response(request)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 145, in inner
    return func(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/module_render.py", line 846, in handle_xblock_callback_noauth
    return _invoke_xblock_handler(request, course_id, usage_id, handler, suffix, course=course)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/module_render.py", line 984, in _invoke_xblock_handler
    resp = instance.handle(handler, req, suffix)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/mixins.py", line 86, in handle
    return self.runtime.handle(self, handler_name, request, suffix)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1334, in handle
    return super(MetricsMixin, self).handle(block, handler_name, request, suffix=suffix)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/runtime.py", line 1025, in handle
    results = handler(request, suffix)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/lti_2_util.py", line 85, in lti_2_0_result_rest_handler
    return self._lti_2_0_result_put_handler(request, real_user)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/lti_2_util.py", line 217, in _lti_2_0_result_put_handler
    self.clear_user_module_score(real_user)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/lti_2_util.py", line 234, in clear_user_module_score
    self.set_user_module_score(user, None, None)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/lti_2_util.py", line 263, in set_user_module_score
    'user_id': user.id,
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/module_render.py", line 477, in publish
    only_if_higher=event.get('only_if_higher'),
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/dispatch/dispatcher.py", line 189, in send
    response = receiver(signal=self, sender=sender, **named)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/grades/signals/handlers.py", line 136, in score_published_handler
    modified=score_modified_time,
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/dispatch/dispatcher.py", line 189, in send
    response = receiver(signal=self, sender=sender, **named)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/grades/signals/handlers.py", line 149, in problem_raw_score_changed_handler
    from traceback import print_stack; print_stack()
ERROR:edx.lti_provider:Outcome Service: Required signal parameter is None. points_possible: None, points_earned: None, user_id: 123, course_id: edx/test_course_a/Test_Course, usa
``

efischer19 commented on lms/djangoapps/courseware/features/lti.feature:
```diff
@@ -110,7 +110,7 @@ Feature: LMS.LTI component
   And the LTI provider deletes my grade and feedback
   And I visit the LTI component (have to reload)
-  Then I see LTI component progress with text "(10.0 points possible)"
+  Then I see LTI component progress with text "(0.0 / 10.0 points)"
```
But somehow, that changes the calculus of what gets displayed on the LTI progress? Does LTI expect the points values to be some weird Heisenbergian `None` value that *can* indicate a reset, but only sometimes?

I'm going to keep investigating and figure out why this PR's changes have affected this test at all, that seems odd to me.

efischer19 commented on lms/djangoapps/grades/tasks.py:
```diff
@@ -41,67 +63,83 @@ def recalculate_subsection_grade(
+        # score should be None only if it was deleted.
+        # Otherwise, it hasn't yet been saved.
+        return score_deleted
+    elif score.module_type == 'openassessment':
```
Changing this check from `score.grade is None` fixed my lettuce tests. LTI "helpfully" issues a score reset by sending a grade of `None, None`, which was triggering the check I had implemented before. Naturally, this failed quickly after as there was no submissions score to check. This check is much more accurate and will avoid the problem of non-openassessment problems hitting this code.

efischer19 commented on lms/djangoapps/grades/signals/handlers.py:
```diff
@@ -56,13 +58,14 @@ def submissions_score_set_handler(sender, **kwargs):  # pylint: disable=unused-a
         return
 
-    PROBLEM_WEIGHTED_SCORE_CHANGED.send(
+    PROBLEM_WEIGHTED_SCORE_CHANGED.send_robust(
```
This change is the other that helped with my lettuce test issues - `send_robust` will send the signal to all handlers of a given signal, regardless of whether there was an exception in an earlier handler. Does anyone object to this change here? It seems smart, given that otherwise we could run into an issue where a poorly-written signal handler gets run first and crashes at a point we aren't expecting.

sanfordstudent replied:
If this PR can go without this change, I would support putting it off until next week when we can do a deep dive on what signal handlers this affects and what test coverage looks like for those handlers. I don't like the idea of only finding problems with LTI because of a change we made in an adjacent app unless the problems arise via a legitimate integration test (as opposed to a test that relies upon a signal that incidentally also gets handled in LTI). That said, if the alternative is to send signals to apps without any integration tests in place, then I think that keeping the `send`s non-robust for now does make sense.

efischer19 replied:
I've changed these back to just `send`

efischer19 commented:
jenkins run all

sanfordstudent commented:
My 👍 still applies.

efischer19 commented:
Alright, the only thing left here before merge is another round of jenkins tests, which I expect (*knocks on wood*) to pass.

No changes to reviewed code other than the suggested switch to checking `score.module_type == 'openassessment'` I called out above.

efischer19 commented:
jenkins run bokchoy

efischer19 commented:
@nasthagiri Did you want to review this once more before it merges? I've left the `_v2` suffix on our method signature per our earlier discussion.
